### PR TITLE
Mobilogram datamodel, detection, deconvolution prototype

### PR DESCRIPTION
**Pull request by SteffenHeu**
Also contains additions of #109

### CODE REVIEW

**Comment by tomas-pluskal** [comment]
Some conflicts appeared after I merged your other branch

**Comment by SteffenHeu** [comment]
Don't merge this just now, i want to get the XYZ chart working & i currently break the IMS trace plot by that

**Comment by SteffenHeu** [comment]
I'd say this is ready to be reviewed & merged. Ill go trhough added files now & check for missing licenses :)

File: src/main/java/io/github/mzmine/gui/preferences/UnitFormat.java [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+
+  public String format(String label, String unit) {
+    switch(this) {
+      case SQUARE_BRACKET:
+        return label + " [" + unit + "]";
+      case ROUND_BRACKED:
+        return label + " (" + unit + ")";
+      case DIVIDE:
+        return  label + " / " + unit;
+      default:
```
Comment (SteffenHeu) [inline src/main/java/io/github/mzmine/gui/preferences/UnitFormat.java:41]: muhaha

Comment (tomas-pluskal) [reply src/main/java/io/github/mzmine/gui/preferences/UnitFormat.java:41]: 😅😅😅

File: src/main/java/io/github/mzmine/datamodel/Frame.java [hunk 2]
```diff
@@ -106,4 +107,10 @@ default double getPrecursorMZ() {
   default int getPrecursorCharge() {
     return 0;
   }
+
+  public ImmutableList<Mobilogram> getMobilograms();
```
Comment (tomas-pluskal) [inline src/main/java/io/github/mzmine/datamodel/Frame.java:111]: If the Mobilogram interface is part of the core data model, you should move it to the datamodel package.

File: src/main/java/io/github/mzmine/datamodel/IMSRawDataFile.java [hunk 3]
```diff
@@ -19,6 +19,8 @@
 package io.github.mzmine.datamodel;
 
 import com.google.common.collect.Range;
+import io.github.mzmine.modules.dataprocessing.featdet_mobilogrambuilder.MobilityDataPoint;
```
Comment (tomas-pluskal) [inline src/main/java/io/github/mzmine/datamodel/IMSRawDataFile.java:22]: MobilityDataPoint should also be moved to datamodel

File: src/main/java/io/github/mzmine/gui/chartbasics/simplechart/SimpleXYZScatterPlot.java [hunk 4]
```diff
@@ -0,0 +1,402 @@
...
+  @Override
+  public ObjectProperty<PlotCursorPosition> cursorPositionProperty() {
+    return cursorPositionProperty;
+  }
+
+  /**
+   * Listens to clicks in the chromatogram plot and updates the selected raw data file accordingly.
+   */
+  private void initializeMouseListener() {
+    getCanvas().addChartMouseListener(new ChartMouseListenerFX() {
```
Comment (tomas-pluskal) [inline src/main/java/io/github/mzmine/gui/chartbasics/simplechart/SimpleXYZScatterPlot.java:185]: Use lambda notation instead?

File: src/main/java/io/github/mzmine/gui/chartbasics/simplechart/SimpleXYZScatterPlot.java [hunk 5]
```diff
@@ -0,0 +1,402 @@
...
+  }
+
+  @Override
+  public void setRangeAxisLabel(String label) {
+    plot.getRangeAxis().setLabel(label);
+  }
+
+  @Override
+  public void setDomainAxisNumberFormatOverride(NumberFormat format) {
+    ((NumberAxis) plot.getDomainAxis()).setNumberFormatOverride(format);
```
Comment (tomas-pluskal) [inline src/main/java/io/github/mzmine/gui/chartbasics/simplechart/SimpleXYZScatterPlot.java:254]: Are the X and Y axes *always* going to be NumberAxis? Cause JFreeChart also supports other types of axes. If we assume they are always NumberAxis, you should make a note about it somewhere in the class Javadocs.

File: src/main/java/io/github/mzmine/gui/chartbasics/simplechart/SimpleXYZScatterPlot.java [hunk 6]
```diff
@@ -0,0 +1,402 @@
...
+  }
+
+  public void setLegendNumberFormatOverride(NumberFormat format) {
+    this.legendAxisFormat = format;
+    onDatasetChanged((XYZDataset) getXYPlot().getDataset());
+  }
+
+  @Override
+  public void addContextMenuItem(String title, EventHandler<ActionEvent> ai) {
+    logger.info("call");
```
Comment (tomas-pluskal) [inline src/main/java/io/github/mzmine/gui/chartbasics/simplechart/SimpleXYZScatterPlot.java:269]: ?

File: src/main/java/io/github/mzmine/gui/chartbasics/simplechart/datasets/ColoredXYZDataset.java [hunk 7]
```diff
@@ -0,0 +1,298 @@
...
+      value = value + delta;
+    }
+
+    return scale;
+  }
+
+  @Override
+  public void run() {
+    status = TaskStatus.PROCESSING;
+    xyValueProvider.computeValues();
```
Comment (tomas-pluskal) [inline src/main/java/io/github/mzmine/gui/chartbasics/simplechart/datasets/ColoredXYZDataset.java:253]: Is the computeValues() function expected to take a long time? Is there any way to cancel it when this task is canceled?

File: src/main/java/io/github/mzmine/gui/chartbasics/simplechart/datasets/ColoredXYZDataset.java [hunk 8]
```diff
@@ -0,0 +1,298 @@
...
+    if (xyzValueProvider.getDomainValues().size() != xyzValueProvider.getRangeValues().size()
+        || xyzValueProvider.getZValues().size() != xyzValueProvider.getRangeValues().size()) {
+      throw new IllegalArgumentException("Number of domain, range or z values does not match.");
+    }
+
+    rangeValues = xyzValueProvider.getRangeValues();
+    domainValues = xyzValueProvider.getDomainValues();
+    zValues = xyzValueProvider.getZValues();
+
+    for (Double rangeValue : rangeValues) {
```
Comment (tomas-pluskal) [inline src/main/java/io/github/mzmine/gui/chartbasics/simplechart/datasets/ColoredXYZDataset.java:267]: You can use Collections.min() and Collections.max() for this

File: src/main/java/io/github/mzmine/gui/chartbasics/simplechart/providers/ExampleXYZProvider.java [hunk 9]
```diff
@@ -0,0 +1,209 @@
...
+          }
+        }
+
+        if (progress != null) {
+          progress.set((double) fi / size);
+        }
+      }
+    } catch (Exception ex) {
+//        logger.log(Level.WARNING, "error in DP", ex);
+      System.out.println("error in DP");
```
Comment (tomas-pluskal) [inline src/main/java/io/github/mzmine/gui/chartbasics/simplechart/providers/ExampleXYZProvider.java:136]: ?

File: src/main/java/io/github/mzmine/gui/mainwindow/MainMenu.fxml [hunk 10]
```diff
@@ -114,7 +114,14 @@
 				userData="io.github.mzmine.modules.dataprocessing.featdet_masscalibration.MassCalibrationModule" />
 		</Menu>
 
-	</Menu>
+    <Menu text="Mobilogram detection">
+      <MenuItem text="Mobilogram builder" onAction="#runModule"
+        userData="io.github.mzmine.modules.dataprocessing.featdet_mobilogrambuilder.MobilogramBuilderModule"/>
+      <MenuItem text="Mobilogram smoother" onAction="#runModule"
```
Comment (tomas-pluskal) [inline src/main/java/io/github/mzmine/gui/mainwindow/MainMenu.fxml:120]: Maybe "Mobilogram smoothing" is better

File: src/main/java/io/github/mzmine/modules/dataprocessing/featdet_mobilogrambuilder/MobilogramBuilderModule.java [hunk 11]
```diff
@@ -0,0 +1,93 @@
...
+  @Nonnull
+  @Override
+  public MZmineModuleCategory getModuleCategory() {
+    return MZmineModuleCategory.FEATUREDETECTION;
+  }
+
+  @Nonnull
+  @Override
+  public String getName() {
+    return "Mobiligram builder";
```
Comment (tomas-pluskal) [inline src/main/java/io/github/mzmine/modules/dataprocessing/featdet_mobilogrambuilder/MobilogramBuilderModule.java:85]: Mobilogram

File: src/main/java/io/github/mzmine/modules/dataprocessing/featdet_mobilogrambuilder/MobilogramBuilderTask.java [hunk 12]
```diff
@@ -0,0 +1,259 @@
...
+//    this.frames = frames;
+    this.frames = (Set<Frame>) scanSelection.getMachtingScans((frames));
+
+    totalFrames = (this.frames.size() != 0) ? this.frames.size() : 1;
+    setStatus(TaskStatus.WAITING);
+  }
+
+  @Override
+  public String getTaskDescription() {
+    return "Detecting mobilograms for frames. " + processedFrames + "/" + totalFrames;
```
Comment (tomas-pluskal) [inline src/main/java/io/github/mzmine/modules/dataprocessing/featdet_mobilogrambuilder/MobilogramBuilderTask.java:74]: Remove the dot after frames?

File: src/main/java/io/github/mzmine/modules/dataprocessing/featdet_mobilogramsmoothing/MobilogramSmootherSetupDialog.java [hunk 13]
```diff
@@ -0,0 +1,163 @@
...
+
+    Collection<? extends Collection<Integer>> resolved = resolver
+        .resolveToIndices(mobilities, intensities, indices);
+
+    Set<Mobilogram> resolvedMobilogram = new HashSet<>();
+    List<MobilityDataPoint> originalDataPoints =
+        originalMobilogram.getDataPoints().stream()
+            .sorted(Comparator.comparingDouble(MobilityDataPoint::getMobility)).collect(
+            Collectors.toList());
+    logger.info("-----");
```
Comment (tomas-pluskal) [inline src/main/java/io/github/mzmine/modules/dataprocessing/featdet_mobilogramsmoothing/MobilogramSmootherSetupDialog.java:138]: ?

File: src/main/java/io/github/mzmine/modules/dataprocessing/featdet_mobilogramsmoothing/MobilogramSmootherSetupDialog.java [hunk 14]
```diff
@@ -0,0 +1,163 @@
...
+    for (Collection<Integer> indicesSet : resolved) {
+      logger.info("Number of deconvolved mobilograms size: " + indicesSet.size());
```
Comment (tomas-pluskal) [inline src/main/java/io/github/mzmine/modules/dataprocessing/featdet_mobilogramsmoothing/MobilogramSmootherSetupDialog.java:140]: deconvolved -> deconvoluted

File: src/main/java/io/github/mzmine/modules/visualization/ims/imsvisualizer/RetentionTimeMobilityHeatMapPlot.java [hunk 15]
```diff
@@ -45,7 +45,7 @@
   private final XYPlot plot;
   static final Font legendFont = new Font("SansSerif", Font.PLAIN, 12);
   private PaintScaleLegend legend;
-  public XYBlockPixelSizeRenderer pixelRenderer;
+  public XYBlockPixelSizeRenderer pixelRenderer; // What does this renderer do? ~SteffenHeu
```
Comment (tomas-pluskal) [inline src/main/java/io/github/mzmine/modules/visualization/ims/imsvisualizer/RetentionTimeMobilityHeatMapPlot.java:48]: Who wrote this comment? :)

File: src/main/java/io/github/mzmine/modules/visualization/mobilogram/MobilogramVisualizerController.java [hunk 16]
```diff
@@ -0,0 +1,255 @@
...
+      public Mobilogram fromString(String string) {
+        return null;
+      }
+    });
+    mobilogramSelector.setCellFactory(listViewListCellCallback);
+  }
+
+  private void initFrameBox() {
+    Callback<ListView<Frame>, ListCell<Frame>> listViewListCellCallback =
+        new Callback<>() {
```
Comment (tomas-pluskal) [inline src/main/java/io/github/mzmine/modules/visualization/mobilogram/MobilogramVisualizerController.java:195]: Could use the lambda notation

File: src/main/java/io/github/mzmine/modules/visualization/mobilogram/MobilogramVisualizerModule.java [hunk 17]
```diff
@@ -0,0 +1,75 @@
...
+import java.util.Collection;
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+
+public class MobilogramVisualizerModule implements MZmineRunnableModule {
+
+  @Nonnull
+  @Override
+  public String getDescription() {
+    return "Visualizes mobilograms of frames";
```
Comment (tomas-pluskal) [inline src/main/java/io/github/mzmine/modules/visualization/mobilogram/MobilogramVisualizerModule.java:40]: How about "Visualizes mobilograms of ion mobility frames"?

File: src/main/java/io/github/mzmine/project/impl/MobilityDataPointStorage.java [hunk 18]
```diff
@@ -0,0 +1,184 @@
...
+    try {
+      FileChannel fileChannel = dataPointsFile.getChannel();
+      fileChannel.lock();
+    } catch (IOException e) {
+      logger.log(Level.WARNING, "Failed to lock the file " + dataPointsFileName, e);
+    }
+
+    // Unfortunately, deleteOnExit() doesn't work on Windows, see JDK
+    // bug #4171239. We will try to remove the temporary files in a
+    // shutdown hook registered in the main.ShutDownHook class
```
Comment (tomas-pluskal) [inline src/main/java/io/github/mzmine/project/impl/MobilityDataPointStorage.java:83]: Did you update the ShutDownHook class to remove the .mobilogram files?

File: src/main/java/io/github/mzmine/parameters/parametertypes/tolerances/mobilitytolerance/MobilityToleranceComponent.java [hunk 19]
```diff
@@ -0,0 +1,73 @@
...
+/**
+ *
+ */
+public class MobilityToleranceComponent extends BorderPane {
+
+  // the same order that the unit enum in RTTolerance is defined in
+  private static final ObservableList<MobilityType> toleranceTypes =
+      FXCollections.observableArrayList(Arrays.asList(MobilityType.values()));
+  private final TextField toleranceField;
+  private final ComboBox<MobilityType> toleranceType;
```
Comment (tomas-pluskal) [inline src/main/java/io/github/mzmine/parameters/parametertypes/tolerances/mobilitytolerance/MobilityToleranceComponent.java:38]: What is the purpose of the user selecting the MobilityType here? What if the user selects MobilityType.NONE?

File: src/main/java/io/github/mzmine/gui/chartbasics/simplechart/SimpleXYZScatterPlot.java (hunk 4 above)
Comment (SteffenHeu) [reply src/main/java/io/github/mzmine/gui/chartbasics/simplechart/SimpleXYZScatterPlot.java:185]: ChartMouseListenerFX is not a functional interface as it has two methods (mouse clicked and mouse moved)

File: src/main/java/io/github/mzmine/gui/chartbasics/simplechart/datasets/ColoredXYZDataset.java (hunk 7 above)
Comment (SteffenHeu) [reply src/main/java/io/github/mzmine/gui/chartbasics/simplechart/datasets/ColoredXYZDataset.java:253]: That depends on the dataset of course. Currently not, but there might be some application in the future. I could pass the TaskStatus of the dataset thread, so the implementing datasets can check that object.

File: src/main/java/io/github/mzmine/project/impl/MobilityDataPointStorage.java (hunk 18 above)
Comment (SteffenHeu) [reply src/main/java/io/github/mzmine/project/impl/MobilityDataPointStorage.java:83]: https://github.com/SteffenHeu/mzmine3/blob/cb7f749527e1735f1c8af30cf9ef214ecda7d940/src/main/java/io/github/mzmine/main/TmpFileCleanup.java#L42 yup ;)

**Comment by tomas-pluskal** [comment]
I think this is not going to work:
`public void computeValues(TaskStatus status)`
because the status is passed by value, not by reference.

**Pull request merged**